### PR TITLE
[RNMobile] Upgrade React Native to version 0.71.15

### PR DESCRIPTION
**Pull request by fluiddot**
**Related PRs:**
* https://github.com/wordpress-mobile/gutenberg-mobile/pull/6522
* https://github.com/wordpress-mobile/WordPress-Android/pull/19964
* https://github.com/wordpress-mobile/WordPress-iOS/pull/22410

## What?

Upgrade React Native to version 0.71.15.

## Why?

* To provide a long-term solution to the workaround introduced in https://github.com/WordPress/gutenberg/pull/57658.
* [Enable upgrading Android SDK to version 14](https://github.com/wordpress-mobile/gutenberg-mobile/issues/6533).

## How?

* Follow [the upgrade helper](https://react-native-community.github.io/upgrade-helper/?from=0.71.11&to=0.71.15).
* Update `Podfile.lock` by running `npm run native preios`.
* Update `react-native` patch file.

> [!WARNING]
> Usually, in React Native upgrades we also update the version in [`react-native-libraries-publisher`](https://github.com/wordpress-mobile/react-native-libraries-publisher) to generate new Android binaries. However, since this upgrade is for a patch version, we assume all third-party libraries support any patch version of `0.71`. For this reason, this step has been skipped.

### Address build failure related to `RCT-Folly`

This is related to the issue shared in https://github.com/WordPress/gutenberg/pull/57667#issuecomment-1882953701. I investigated the problem further and found a workaround.

For some reason, `RCT-Folly` [redefines `clockid_t`](https://github.com/facebook/folly/blob/4baba28200d7446c870e96f3cdbeb492f54625d0/folly/portability/Time.h#L33), however, this seems not needed in newer iOS SDK versions. This was been fixed in `RCT-Folly` version (`v2022.08.29.00` - https://github.com/facebook/folly/commit/4a2410fae65afb85e1fec6d922005054b05de59f) but current React Native version points to an older version ([`2021.07.22.00` - reference](https://github.com/facebook/react-native/blob/5761ad6437b29dc559fce0914f2a8e71d02d25b3/third-party-podspecs/RCT-Folly.podspec#L6)). The version has been bumped in React Native version 0.73.0 ([`2022.05.16.00` - reference](https://github.com/facebook/react-native/blob/1eb4bf0d3af1fe267f5de6bf4bbe1e65d41fea1b/packages/react-native/third-party-podspecs/RCT-Folly.podspec#L6)), so we'd need to keep the patch until we upgrade.

Regarding the patch, this is the part we need from `react_native_pods` script ([reference](https://github.com/facebook/react-native/blob/5761ad6437b29dc559fce0914f2a8e71d02d25b3/scripts/react_native_pods.rb#L359-L364)) that removes the condition related to checking the minimum version required ([reference](https://github.com/facebook/folly/blob/4baba28200d7446c870e96f3cdbeb492f54625d0/folly/portability/Time.h#L33)):

## Testing Instructions

> [!NOTE]
> It's recommended to use the following installable builds for testing:
> * [Android](https://github.com/wordpress-mobile/WordPress-Android/pull/19964#issuecomment-1895427275)
> * [iOS](https://github.com/wordpress-mobile/WordPress-iOS/pull/22410#issuecomment-1895420518)

Nothing in particular, it should be enough by smoke testing the editor.

### Testing Instructions for Keyboard

N/A

## Screenshots or screencast 
N/A

### CODE REVIEW

**Comment by fluiddot** [comment]
⚠️  The demo app project can't be built due to this error ([reference](https://github.com/WordPress/gutenberg/actions/runs/7460562048/job/20298872238?pr=57667#step:12:7924)):
```
Error: typedef redefinition with different types ('uint8_t' (aka 'unsigned char') vs 'enum clockid_t')
typedef uint8_t clockid_t;
```

I updated the `Podfile` with the solution referenced in https://github.com/facebook/react-native/issues/33178#issuecomment-1737079962, but still doesn't work. We'd need to investigate this further to finish the upgrade, as well as future upgrades.

**Comment by paulmbw** [comment]
Hi, I noticed you're having issues with upgrading React Native. I've managed to upgrade my [app](https://launchtoday.dev/) to 0.72.3 with some effort. I posted a tweet [here](https://twitter.com/paulwaweruco/status/1744838049748365689) about it, and happy to share the git diff of my changes if that helps (I can't post it here as it's a paid product but I'm happy to walk you through the code on a call at no cost)

**Comment by fluiddot** [comment]
> Hi, I noticed you're having issues with upgrading React Native. I've managed to upgrade my [app](https://launchtoday.dev/) to 0.72.3 with some effort. I posted a tweet [here](https://twitter.com/paulwaweruco/status/1744838049748365689) about it, and happy to share the git diff of my changes if that helps (I can't post it here as it's a paid product but I'm happy to walk you through the code on a call at no cost)

Thank you, @paulmbw, for sharing this valuable information and for your willingness to assist 🙇. The integration of React Native into the Gutenberg project exhibits a slightly different and more complex structure compared to a typical React Native app. Throughout our previous upgrades, we have encountered various blockers, and it is likely that the issue mentioned in [this comment](https://github.com/WordPress/gutenberg/pull/57667#issuecomment-1882953701) is related to the unique integration of React Native.

My purpose with this PR was mainly to test the waters and check if we could quickly apply the patch version. However, as shared above, it appears that this is not the case. Hence, we've decided to temporarily halt our efforts and plan to revisit this matter upon upgrading to version 0.72.3.

**Comment by fluiddot** [comment]
> I updated the `Podfile` with the solution referenced in [facebook/react-native#33178 (comment)](https://github.com/facebook/react-native/issues/33178#issuecomment-1737079962), but still doesn't work. We'd need to investigate this further to finish the upgrade, as well as future upgrades.

I investigated the issue further and managed to apply a workaround in https://github.com/WordPress/gutenberg/pull/57667/commits/b72f56af7d31944edef87a7c6565eaa665851620. More details can be found in the PR's description.

File: packages/react-native-editor/ios/GutenbergDemo.xcodeproj/project.pbxproj
Comment (fluiddot) [inline packages/react-native-editor/ios/GutenbergDemo.xcodeproj/project.pbxproj:1]: These changes have been introduced in [React Native version 0.71.14](https://github.com/facebook/react-native/blob/main/CHANGELOG.md#v07114) as part of an Xcode 15 patch.

**Review by geriux — APPROVED** [review]
LGTM! I've smoke tested the editor on both platforms, I've also checked that the accessibility patch works as expected 🚀

**Comment by fluiddot** [comment]
Heads up that I've just added the following warning to the PR's description:
> [!WARNING]
> Usually, in React Native upgrades we also update the version in [`react-native-libraries-publisher`](https://github.com/wordpress-mobile/react-native-libraries-publisher) to generate new Android binaries. However, since this upgrade is for a patch version, we assume all third-party libraries support any patch version of `0.71`. For this reason, this step has been skipped.

**Pull request merged**